### ISSUE TITLE
Question about loading pretrained model

### ISSUE THREAD

**Issue by LIU-FAYANG** [ultralytics/yolov5#3494]
## ❔Question
HI, thanks for your great work. I tried to load the pretrianed model with torch.hub.load() and it kept installing opencv-python, which is not desired because the object tracker I'm using only support opencv-contrib-python. Is there any method to avoid installing opencv-python when loading the pretrained models?

Also, I tied use  detector = torch.load('yolov5s.pt',map_location=torch.device('cpu'))['model'] to load the model locally and some error showed up: do you have any suggestions to solve this?
Thanks for your kind help!:)

Traceback (most recent call last):
  File "loitering_main.py", line 188, in main
    dets_ = detector(imgs=frame[:,:,::-1])  # reverse the channels to RGB format for torch model since opencv is in BGR format
  File "/home/psl/anaconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
TypeError: forward() got an unexpected keyword argument 'imgs'

**Comment by glenn-jocher** [comment]
@LIU-FAYANG good news 😃! Your original issue may now be fixed ✅ in PR #3495. This PR adds opencv-python to the hubconf.py check_requirements() `exclude` list, allowing for 3rd party or contrib versions of OpenCV to be used. To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by LIU-FAYANG** [comment]
Hi, I tried to git clone again in current repository and change the  code to:
detector = torch.hub.load('ultralytics/yolov5', 'yolov5s',  force_reload = True) and the terminal still shows:

requirements: opencv-python>=4.1.2 not found and is required by YOLOV5, attempting auto-update.

**Comment by glenn-jocher** [comment]
@LIU-FAYANG hi, I'm sorry about that. Yes I think my update only propagated to `develop` branch when PyTorch Hub is pulling `master` branch by default. I'll push an identical update to `master` shortly which should fix this for you.

@SkalskiP @KalenMike this example issue.

**Comment by SkalskiP** [comment]
@glenn-jocher Can I get some more information on why this is happening?

**Comment by glenn-jocher** [comment]
@SkalskiP Under the hood the torch.hub.load() command is cloning the repo that's passed in the first argument, i.e. 'ultralytics/yolov5', searching for hubconf.py at the top level (which is why we can't place it in the yolov5/ dir in the pip package), and then inside hubconf.py finding a function called 'yolov5s()', which it expects to return a model from.

```python
model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)
```

The problem seems to be that torch hub is pulling master branch by default. So my fix in #3495 doesn't make it into these hub models since I merged it into develop branch.

**Comment by glenn-jocher** [comment]
@SkalskiP the documentation explains a bit more here:
https://pytorch.org/docs/stable/hub.html#loading-models-from-hub

<img width="865" alt="Screenshot 2021-06-07 at 12 10 38" src="https://user-images.githubusercontent.com/26833433/120999432-6393a700-c789-11eb-8907-3eb6915ea836.png">

**Comment by glenn-jocher** [comment]
This is correct. All images are used to compute mAP for each class, not just the images containing the classes.

<img width="1391" alt="Screenshot 2021-06-07 at 12 23 03" src="https://user-images.githubusercontent.com/26833433/121001003-20d2ce80-c78b-11eb-8391-7ad3accc7006.png">

**Comment by SkalskiP** [comment]
@glenn-jocher and why it is wrong that the torchhub is integrated with the `master` branch? `master` is a stable branch.

**Comment by glenn-jocher** [comment]
@SkalskiP I've fixed this problem in #3495, pushed to develop.

@LIU-FAYANG can't receive this fix because PyTorch Hub uses master.

**Comment by glenn-jocher** [comment]
@SkalskiP PyTorch Hub is not in the wrong, we are with our branch structure.

**Comment by SkalskiP** [comment]
@glenn-jocher so you merged hot-fix to `develop` not to `master`? To me "branch structure" is correct. It is simply that we should merge fixes to master and features to develop.

**Comment by glenn-jocher** [comment]
@SkalskiP if we merge fixes to master then everyone who git clones our repo, i.e. 2000 people a day, will not have access to the latest fixes, and will continue to raise bug reports since you've placed develop as our default branch.

**Comment by SkalskiP** [comment]
@glenn-jocher a, now I understand. So I did set `develop` as the default branch only to make sure that it will be set as PR target branch automatically. Setting `develop` as default is not mandatory in any way. It was only for convenience. If we feel it causes more problems let's set `master` as default, but then we need to remember to double-check the PR target branch. But I guess in that case that makes much more sense. So my proposal is:

* Keep `master` <- `develop` <- `feature_branch` flow
* Set `master` as the default branch so that it would load after `git clone`
* Remember to set correct target branch when someone create PR
* Merge fixes into `master` and than rebase develop on fresh `master`
* Merge new features into `develop`

@glenn-jocher what do you think about that?

**Comment by glenn-jocher** [comment]
@SkalskiP yeah maybe that's a better route, though there might be confusion indicating to the PR submitters which branch they should rebase against.

Let's discuss all the options at our meeting tomorrow with Kalen, we'll try to figure out a solution there. Thanks!

**Comment by glenn-jocher** [comment]
@LIU-FAYANG good news 😃! Your original issue may now be fixed ✅ in PR #3507 which I just merged to `master`. This PR is a duplicate PR of #3495 which was merged to `develop` previously, which PyTorch Hub does not use, and which was the reason my first fix attempt failed to resolve your issue. To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
